pkg/api/validate/dynamic/workersku.go

```diff
@@ -0,0 +1,55 @@
+package dynamic
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/computeskus"
+)
+
+// ValidateWorkerSku uses resourceSkusClient to ensure that the VMSize listed in the cluster document WorkerProfile is available for use in the target region.
+func (dv *dynamic) ValidateWorkerSku(ctx context.Context, location string, subscriptionID string, oc *api.OpenShiftCluster) error {
+	// Get a list of available worker SKUs, filtering by location.
+	filter := fmt.Sprintf("location eq %s", location)
+	skus, err := dv.resourceSkusClient.List(ctx, filter)
+	if err != nil {
+		return err
+	}
+
+	sortedSkus := computeskus.FilterVmSizes(skus, location)
+
+	// In case there are multiple WorkerProfiles listed in the cluster document (such as post-install),
+	// compare VMSize in each WorkerProfile to the resourceSkusClient call above to ensure that the sku is available in region.
+	for _, workerprofile := range oc.Properties.WorkerProfiles {
+		skuAvailable := false
+		workerProfileSku := string(workerprofile.VMSize)
+
+		for name, sku := range sortedSkus {
+			if name == workerProfileSku {
+				skuAvailable = true
+
+				// Fail if sku is available, but restricted within the subscription. Restrictions are subscription-specific.
+				// https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/error-sku-not-available
+				for _, restriction := range *sku.Restrictions {
+					restrictedLocations := strings.Join(*restriction.RestrictionInfo.Locations, " ")
+
+					if strings.Contains(restrictedLocations, location) && skuAvailable {
+						return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "", "The selected sku %v is restricted in region %v for subscription %v: restricted locations contain %v", workerProfileSku, location, subscriptionID, restrictedLocations)
+					}
+				}
```

Comment:
I'm not totally sure yet, but on first pass I could imagine a false positive if the restriction `type` is something other than `Location` or `Zone`:
```
      "restrictions": [
        {
          "reasonCode": "NotAvailableForSubscription",
          "restrictionInfo": {
            "locations": [
              "japanwest"
            ]
          },
          "type": "Location",
          "values": [
            "japanwest"
          ]
        }
      ],
```

Reply:
@m1kola there would also be a false positive if `location` was equal to `" "`. Am I close?

Reply:
@m1kola I think I've got it - we would get a false positive if the restricted location is something like `eastus2`, and `location` is `eastus`.

So rather than do a `strings.Contains` comparison, we need to do an exact comparison.